spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestChangelogReader.java

```diff
@@ -191,39 +214,361 @@ public void testMixDeleteAndInsert() throws IOException {
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId1, 0, records1);
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId2, 1, records2);
+
+    assertEquals("Should have expected rows", expectedRows, internalRowsToJava(rows));
+  }
+
+  @Test
+  public void testFlinkScenario1() throws IOException {
+    List<Record> rl1 = Lists.newArrayList();
+    GenericRecord record = GenericRecord.create(table2.schema());
+    rl1.add(record.copy("id", 1, "data", "a"));
+    rl1.add(record.copy("id", 2, "data", "b"));
+    rl1.add(record.copy("id", 3, "data", "c"));
+    DataFile df1 = writeDataFile(rl1);
+    table2.newAppend().appendFile(df1).commit();
+    long snapshotId1 = table2.currentSnapshot().snapshotId();
+
+    // update row with id=2 by deleting it using an equality delete and writing a new row with id=2
+    Schema deleteRowSchema = table2.schema().select("id");
+    Record idDelete = GenericRecord.create(deleteRowSchema);
+    List<Record> idDeletes = Lists.newArrayList(idDelete.copy("id", 2));
+
+    DeleteFile eqDeletes =
+        FileHelpers.writeDeleteFile(
+            table2,
+            Files.localOutput(File.createTempFile("junit", null, temp.toFile())),
+            TestHelpers.Row.of(0),
+            idDeletes,
+            deleteRowSchema);
+
+    List<Record> rl2 = Lists.newArrayList(record.copy("id", 2, "data", "bb"));
+    DataFile df2 = writeDataFile(rl2);
+
+    table2.newRowDelta().addDeletes(eqDeletes).addRows(df2).commit();
+    long snapshotId2 = table2.currentSnapshot().snapshotId();
+
+    // update row with id=2 again the same way
+    List<Record> idDeletes2 = Lists.newArrayList(idDelete.copy("id", 2));
+
+    DeleteFile eqDeletes2 =
+        FileHelpers.writeDeleteFile(
+            table2,
+            Files.localOutput(File.createTempFile("junit", null, temp.toFile())),
+            TestHelpers.Row.of(0),
+            idDeletes2,
+            deleteRowSchema);
+
+    List<Record> rl3 = Lists.newArrayList(record.copy("id", 2, "data", "bbb"));
+    DataFile df3 = writeDataFile(rl3);
+
+    table2.newRowDelta().addDeletes(eqDeletes2).addRows(df3).commit();
+    long snapshotId3 = table2.currentSnapshot().snapshotId();
+
+    List<InternalRow> rows = getChangelogRows(table2);
+
+    List<Record> deleted1 = Lists.newArrayList(record.copy("id", 2, "data", "b"));
+    List<Record> deleted2 = Lists.newArrayList(record.copy("id", 2, "data", "bb"));
+
+    List<Object[]> expectedRows = Lists.newArrayList();
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId1, 0, rl1);
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId2, 1, deleted1);
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId2, 1, rl2);
+    // in snapshot 3, eqDeletes2 applies to both df1 and df2, so there are two deletes emitted
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId3, 2, deleted1);
```

Comment:
I think this one is not correct. This record is not deleted in snapshot3

Reply:
I have fixed the expected results to remove this row.